### PR TITLE
Add support for `{date}` replacement

### LINKED ISSUE #30: Date placeholder/replacement
Hey fregante,

I would like to request to add **date placeholder/replacement of commit**. You could add this placeholder – `{date}` to `commit-template` where it will "print" date of commit and adds link to commit on it. You could add it like follows:

```yaml
...
- name: Create release with date
  uses: fregante/release-with-changelog@v3
  with:
    token: ${{ secrets.GITHUB_TOKEN }}
    commit-template: '- {date} {title}'
```

And it will produce release with text `- [4.2.2021](https://github.com/fregante/release-with-changelog/commit/6e89f5c9d6622f54ca284c06aa89fa7391cd86f7) Update user in examples`:

- [4.2.2021](https://github.com/fregante/release-with-changelog/commit/6e89f5c9d6622f54ca284c06aa89fa7391cd86f7) Update user in examples

It's better to have date (pointing to commit) instead of commit hash in changelog or in mail notification.

### PR DESCRIPTION
**Pull request by BelKed**
Resolves #30.

- [x] Added date replacement
- [x] Better formatting in actions.yml + Added option for date format
- [x] Updated readme.md

### CODE REVIEW

**Comment by fregante** [comment]
There are a lot of unrelated changes (the action.yml formatting isn’t “better” and is unrelated) and the file in `distribution` should not be modified manually. The source is `/index.js`. Please ensure that exclusively changes related to dates are included in this PR.

**Comment by BelKed** [comment]
- [x] Oh, I will move edits to `/index.js`. I wondered why `index.js` in `distribution` had so many lines – I think, that was file, I should edit.
  - Also, shouldn't be it in `.gitignore` and builded and committed by Github Action on push/pull request?
- [x] And I have to fix bug, where longer or shorter dates caused bad date, hash and title of commit in description.

---

- There should be updated documentation with building/contributing/developing...

File: action.yml [hunk 1]
```diff
@@ -1,5 +1,4 @@
 name: 'Release with Changelog'
-icon: 'tag'
```
Comment (fregante) [inline action.yml:2]: Unrelated change

File: generate-release-notes.js [hunk 2]
```diff
@@ -9,13 +9,15 @@ async function generateReleaseNotes({
...
 	exclude = '',
 	commitTemplate = '- {hash} {title}',
-	releaseTemplate = '{commits}\n\n{range}'
+	releaseTemplate = '{commits}\n\n{range}',
+	dateFormat = '--date=short'
 }) {
 	// Get commits between computed range
-	let {stdout: commits} = await execFile('git', ['log', '--format=%H%s', range]);
+	let {stdout: commits} = await execFile('git', ['log', '--format=%H¬%ad¬%s', dateFormat, range]);
 	commits = commits.split('\n').filter(Boolean).map(line => ({
```
Comment (fregante) [inline generate-release-notes.js]: Let’s only call split once.

```suggestion
	commits = commits.split('\n').filter(Boolean).map(line => {
	    const [hash, date, title] = line.split(etc)
```

File: generate-release-notes.js [hunk 3]
```diff
@@ -28,11 +30,12 @@ async function generateReleaseNotes({
...
 		commitEntries.push('_Maintenance release_');
 	} else {
-		for (const {hash, title} of commits) {
+		for (const {hash, date, title} of commits) {
 			const line = commitTemplate
 				.replace('{hash}', hash)
 				.replace('{title}', title)
-				.replace('{url}', repoURL + '/commit/' + hash);
+				.replace('{url}', repoURL + '/commit/' + hash)
+				.replace('{date}', '[' + date + '](' + repoURL + '/commit/' + hash + ')');
```
Comment (fregante) [inline generate-release-notes.js]:
```suggestion
				.replace('{date}', date);
```

The link is unrelated. If the user wants a linked date they can use `[{date}]({url})` or whatever format they please

File: readme.md [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# release-with-changelog
```
Comment (fregante) [inline readme.md]: Unrelated

File: readme.md [hunk 5]
```diff
@@ -91,6 +92,14 @@ Default: _latest tag available_
 
 Specific tag to generate changelog against.
 
+### date-format
+
+Default: `short` <br>
+Example: `local` <br>
+Example: `%d.%m.%Y`
+
+Format of `{date}` replacement. More info can be found in [StackOverflow answer](https://stackoverflow.com/questions/7853332/#34778736).
```
Comment (fregante) [inline readme.md]:
```suggestion
Format of `{date}` replacement as used by [git log --date](https://git-scm.com/docs/git-log#Documentation/git-log.txt---dateltformatgt), as either a keyword or a `strftime` string. More info can be found in [StackOverflow answer](https://stackoverflow.com/questions/7853332/#34778736).
```

File: action.yml (hunk 1 above)
Comment (BelKed) [reply action.yml:2]: I removed it because it causes error – _Property icon is not allowed._
So, should I put it back?

File: readme.md (hunk 5 above)
Comment (BelKed) [reply readme.md]: OK, I add it as you wants.

File: readme.md (hunk 4 above)
Comment (BelKed) [reply readme.md]: OK, I put it back, as it was.

File: generate-release-notes.js (hunk 3 above)
Comment (BelKed) [reply generate-release-notes.js]: OK, I add it as you wants. Also, I rewrite it in readme.

File: generate-release-notes.js (hunk 2 above)
Comment (BelKed) [reply generate-release-notes.js]: I tried it and it causes error – `TypeError: Right side of assignment cannot be destructured`
```js
commits = commits.split('\n').filter(Boolean).map(line => {
	var [hash, date, title] = line.split('¬'),
	hash = hash.slice(0, 8)
});
```
I am newbie to JS and I don't know where is problem...

Comment (fregante) [reply generate-release-notes.js]: You’re using a comma at the end of the line but it should be a semicolon

File: action.yml (hunk 1 above)
Comment (fregante) [reply action.yml:2]: You’re right. It’s a duplicate of branding.icon. It’s fine as is

Comment (BelKed) [reply action.yml:2]: OK

**Comment by fregante** [comment]
> shouldn't be it in .gitignore and builded and committed by Github Action

Unfortunately this doesn’t work with GitHub Actions. They want a pre-built file committed to the repo. 

Also no need to force push. Just add commits normally.

File: generate-release-notes.js [hunk 6]
```diff
@@ -9,14 +9,15 @@ async function generateReleaseNotes({
...
 	// Get commits between computed range
-	let {stdout: commits} = await execFile('git', ['log', '--format=%H%s', range]);
-	commits = commits.split('\n').filter(Boolean).map(line => ({
-		hash: line.slice(0, 8),
-		title: line.slice(40)
-	}));
+	let {stdout: commits} = await execFile('git', ['log', '--format=%H¬%ad¬%s', dateFormat, range]);
+	commits = commits.split('\n').filter(Boolean).map(line => {
+		var [hash, date, title] = line.split('¬');
+		hash = hash.slice(0, 8)
```
Comment (fregante) [inline generate-release-notes.js]:
```suggestion
		return {
		hash: hash.slice(0, 8), 
		date, 
		title
		};
```

File: index.js [hunk 7]
```diff
@@ -11,6 +11,8 @@ async function run() {
 		const releaseTemplate = core.getInput('template');
 		const commitTemplate = core.getInput('commit-template');
 		const exclude = core.getInput('exclude');
+		let dateFormat = core.getInput('date-format');
+		dateFormat = ['relative', 'local', 'default', 'iso', 'rfc', 'short', 'raw'].includes(dateFormat) ? ('--date=' + dateFormat) : ('--date=format:' + dateFormat);
```
Comment (fregante) [inline index.js]: Don’t pass a whole cli flag to the generateReleaseNotes function, you can create the flag inside it. 
```suggestion
		dateFormat = dateFormat.includes('%') ? 'format:' + dateFormat : dateFormat;
```

**Review by fregante — COMMENTED** [review]
Can you also add some tests to ensure it works?

File: generate-release-notes.js (hunk 6 above)
Comment (BelKed) [reply generate-release-notes.js]: Now it works 😄

File: index.js (hunk 7 above)
Comment (BelKed) [reply index.js]: There should be also `--date=` implemented.
So it will be:
```js
dateFormat = dateFormat.includes('%') ? '--date=format:' + dateFormat : '--date=' + dateFormat;
```

**Comment by BelKed** [comment]
> Can you also add some tests to ensure it works?

And how?
I tested it in on my private repo and date option works...

File: index.js (hunk 7 above)
Comment (fregante) [reply index.js]: Read my comment above the suggestion

**Comment by fregante** [comment]
There’s a file that ends with .test.js. Look at how the other features are tested a create similar tests for this new option. It should test for both a keyword and some custom formats

File: index.js (hunk 7 above)
Comment (BelKed) [reply index.js]: Which one do you mean?

**Comment by BelKed** [comment]
I added date test, but after I run it it except a error on `generates changelog with custom date presets`. It's function with test custom date format. I don't know what is wrong with it.

File: index.js (hunk 7 above)
Comment (fregante) [reply index.js]: > Don’t pass a whole cli flag to the generateReleaseNotes function, you can create the flag inside it.

Comment (BelKed) [reply index.js]: OK, I will edit it...

**Review by fregante — APPROVED** [review]
Perfect, thanks! I’ll leave it here a day or 2 and then merge it if no one complains

**Comment by BelKed** [comment]
I added more date formats, especially for the date without leading zero and also for testing in `generate-release-notes.test.js`... It's using [`dateformat`](https://www.npmjs.com/package/dateformat).

**Comment by fregante** [comment]
Please undo, git date formats are good enough.

**Comment by fregante** [comment]
> Also no need to force push. Just add commits normally.

**Comment by BelKed** [comment]
I force pushed, because I didn't want to mess code. I just made small typo and I fixed it.

File: readme.md [hunk 8]
```diff
@@ -91,6 +92,14 @@ Default: _latest tag available_
 
 Specific tag to generate changelog against.
 
+### date-format
+
+Default: `short` <br>
+Example: `local` <br>
+Example: `%d.%m.%Y`
+
+Format of `{date}` replacement as used by [git log --date](https://git-scm.com/docs/git-log#Documentation/git-log.txt---dateltformatgt), as either a keyword or a `strftime` string. More info can be found in [StackOverflow answer](https://stackoverflow.com/questions/7853332/#34778736).
```
Comment (notlmn) [inline readme.md]:
```suggestion
Format of `{date}` replacement as used by [`git log --date`](https://git-scm.com/docs/git-log#Documentation/git-log.txt---dateltformatgt), as either a keyword or a `strftime` string. More info can be found in [StackOverflow answer](https://stackoverflow.com/questions/7853332/#34778736).
```

File: readme.md [hunk 9]
```diff
@@ -74,6 +74,7 @@ Markdown template for each commit entry in release notes. Available replacements
 - `{title}` A single line title of the commit.
 - `{hash}` Abbreviated commit hash, gets linkified automatically in release notes.
 - `{url}` Plain link to commit on GitHub.
+- `{date}` Date of commit, could be specified in [`date-format`](#date-format) below.
```
Comment (notlmn) [inline readme.md]: nit:

```suggestion
- `{date}` Date of commit, could be specified in [`date-format`](#date-format).
```

**Review by notlmn — COMMENTED** [review]
Apart from the minor nits, looks good to me.

File: readme.md (hunk 8 above)
Comment (BelKed) [reply readme.md]: OK

File: readme.md (hunk 9 above)
Comment (BelKed) [reply readme.md]: OK

**Comment by BelKed** [comment]
I remade whole test, because custom date format check is failing. It was failing because `generateReleaseNotes` function does not run `git log` with custom date parameters...

**Comment by fregante** [comment]
Not a bad idea, I dislike mocks for this exact reason, they give a false sense of security.

**Comment by BelKed** [comment]
Could you merge this pull request now, so I can open another, that fixes another issue?

**Comment by fregante** [comment]
Thank you @BelKed!

**Comment by fregante** [comment]
Unfortunately the commit titles aren’t being excluded from the replacement. The commit here included `{date}` and was replaced with the actual date. 

7C1EE901-BF0F-4106-8BEC-C2F460768459
